airbyte-integrations/connectors/source-mailgun/source_mailgun/source.py

```diff
@@ -0,0 +1,154 @@
+        """
+        next_page: Optional[HttpUrl] = response.json().get('paging', {}).get('next')
+        return {"url": next_page} if next_page and self._pre_parse_response(response) else None
+
+    def path(
+            self,
+            stream_state: Mapping[str, Any] = None,
+            stream_slice: Mapping[str, Any] = None,
+            next_page_token: Optional[Mapping[str, HttpUrl]] = None,
+    ) -> str:
+        # TODO: Requires improve URL creation in the CDK by using urllib.parse.urljoin
+        #  (airbyte_cdk.sources.streams.http.http._create_prepared_request)
+        return next_page_token["url"] if next_page_token else ""
+
+    def request_params(
+            self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, any] = None,
+            next_page_token: Optional[Mapping[str, Any]] = None
+    ) -> MutableMapping[str, Any]:
+        return {}
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        """
+        :return an iterable containing each record in the response
+        """
+        yield from self._pre_parse_response(response)
+
+    @staticmethod
+    def _pre_parse_response(response: requests.Response) -> List:
+        return response.json()['items']
+
+
+class Domains(MailgunStream):
+    primary_key = "name"
+
+    def path(self, *args, next_page_token: Optional[Mapping[str, Any]] = None, **kwargs) -> str:
+        # return super().path(*args, **kwargs) or "domains"  # TODO: Requires the CDK update (see MailgunStream.path())
+        path: str = "domains"
+        if next_page_token:
+            token: str = next_page_token['url'].rpartition('/')[2]
+            path = f"{path}/{token}"
+
+        return path
+
+
+# Basic incremental stream
+class IncrementalMailgunStream(MailgunStream, ABC):
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> \
+            Mapping[str, Any]:
+        return {}
+
+
+class Events(IncrementalMailgunStream):
+    # TODO: Event Polling. See https://documentation.mailgun.com/en/latest/api-events.html#event-polling
+
+    cursor_field = "timestamp"
+
+    primary_key = "id"
+
+    def __init__(self, config: Mapping[str, Any], *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.timestamp = config.get('timestamp', 0)
+        self.ascending = config.get('ascending', True)
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> \
+            Mapping[str, Any]:
+        if current_stream_state is not None and 'timestamp' in current_stream_state:
+            current_parsed_timestamp = current_stream_state['timestamp']
+            latest_record_timestamp = latest_record['timestamp']
+            return {'timestamp': max(current_parsed_timestamp, latest_record_timestamp)}
+        else:
+            return {'timestamp': self.timestamp}
```

Comment:
the logic is wrong, it should take value from the record

Reply:
It does.

Reply:
@sergei-solonitcyn 
not really, just trace the logic when there is no state -> the code will return `start_timestamp` instead of returning an actual value from the record.
Also this is a good example why EAFP can't be used everywhere, it is hard to tell which line raise exception, and that will require different `except` logic